font/src/ft/mod.rs

```diff
@@ -529,6 +581,66 @@ impl FreeTypeRasterizer {
     }
 }
 
+fn downsample_bitmap(mut bitmap_glyph: RasterizedGlyph, fixup_factor: f64) -> RasterizedGlyph {
+    // Only scale colored buffers which are bigger than required
+    let bitmap_buffer = match (&bitmap_glyph.buf, fixup_factor.partial_cmp(&1.0)) {
+        (BitmapBuffer::RGBA(buffer), Some(Ordering::Less)) => buffer,
+        _ => return bitmap_glyph,
+    };
+
+    let bitmap_width = bitmap_glyph.width as usize;
+    let bitmap_height = bitmap_glyph.height as usize;
+
+    let target_width = (bitmap_width as f64 * fixup_factor) as usize;
+    let target_height = (bitmap_height as f64 * fixup_factor) as usize;
+
+    let downsampling_step = 1.0 / fixup_factor;
+
+    let mut downsampled_buffer = Vec::<u8>::with_capacity(target_width * target_height * 4);
+
+    for line_index in 0..target_height {
+        let line_index = line_index as f64;
+        let source_line_start = (line_index * downsampling_step).floor() as usize;
+        let source_line_end =
+            min(((line_index + 1.) * downsampling_step).floor() as usize, bitmap_height);
+
+        for column_index in 0..target_width {
+            let column_index = column_index as f64;
+            let source_column_start = (column_index * downsampling_step).floor() as usize;
```

Comment:
It should be `floor` and not `ceil`. Your approach has the bug with missing pixels/etc, just zoom in with your diff and you should crash.